[BE][autograd Function] Raise an error if input is returned as-is and saved for forward or backward in setup_context

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97212

Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc @ezyang @gchanan

soulitzer committed:
[BE][autograd Function] Raise an error if input is returned as-is and saved for backward

[ghstack-poisoned]

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for backward"




cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented:
This one is too bc-breaking. Closing in favor of https://github.com/pytorch/pytorch/pull/97318

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for backward"


Fixes https://github.com/pytorch/pytorch/issues/96887

Updates both the cpp and python custom autograd Functions.

TODO: Since this is bc-breaking maybe I should import this separately to see if there are internal failures.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented on test/cpp/api/autograd.cpp:
```diff
@@ -254,6 +254,28 @@ TEST(AutogradAPITests, AnomalyMode) {
+  Variable y = torch::randn({5, 5}, torch::requires_grad());
+  MyFunction::apply(x, y);
+}
+
```
cpp behavior isn't changed, this test just makes sure we preserve current behavior

zou3519 commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -427,15 +443,21 @@ optional_variable_list _wrap_outputs(
+      dirty_inputs,
+      raw_outputs,
+      cdata,
+      to_save_if_setup_context);
```
As another sanity check: It's not a problem for forward-mode AD if this happens (i.e. there is no ambiguity if we save_for_forward an input that has been returned as an output), right? Should we add a test for that?

soulitzer replied:
I'd need to check the functorch behavior, but either way I'd happy with also disallowing save_for_forward on inputs saved as outputs when used with setup-context as well.

soulitzer replied:
Test added.

soulitzer replied:
Ugh I think we may need to disallow save_for_forward as well
```python
class A(torch.autograd.Function):
    @staticmethod
    def forward(x):
        return x

    @staticmethod
    def setup_context(ctx, inputs, output):
        ctx.save_for_forward(output, inputs[0])

    @staticmethod
    def jvp(ctx, x_t):
        out, inp = ctx.saved_tensors
        # It does matter whether you use out, inp here!
        return x_t * out

x = torch.tensor(2., device=device)
x_t = torch.tensor(2., device=device)

# If we do just jvp, everything is OK
jvp(A.apply, (x,), (x_t,))

def jvp_fn(fn):
    def inner(x, x_t):
        return jvp(fn, (x,), (x_t,))[1]
    return inner

# There is a difference between using inp/out in the grad(jvp case
# If we use inp, the grad( does not need to backprop through any
# ABackward nodes
# If we use out, we would error out here because the out has
# ABackward as its grad_fn and it expects the vjp to be
# implemented 
grad(jvp_fn(A.apply))(x, x_t)
```

soulitzer replied:
It is now updated to do so.

zou3519 replied:
Just to check, grad(vmap( is fine, right? I'd expect that case to be fine because vmap would preserve the identity of the input

soulitzer replied:
If you mean vmap(grad, then yes. (but it doesn't matter that vmap preserves identity because the autograd Function is still applied to the inner-most tensor and that aliases always, so the fix [below ](https://github.com/pytorch/pytorch/pull/97212#discussion_r1148046522) also fixes that.

zou3519 replied:
I was talking about grad(vmap:
```
import torch
from torch.func import grad

class Cube(torch.autograd.Function):
    generate_vmap_rule = True

    @staticmethod
    def forward(x):
        return x ** 3, x

    @staticmethod
    def setup_context(ctx, inputs, outputs):
        cube, x = outputs
        ctx.save_for_backward(x)

    @staticmethod
    def backward(ctx, grad_output, grad_x):
        # NB: grad_x intentionally not used in computation
        x, = ctx.saved_tensors
        result = grad_output * 3 * x ** 2
        return result


def f(x):
    cube, inp = Cube.apply(x)
    return cube

x = torch.tensor([1., 2., 3.], requires_grad=True)

print("============= functorch")
grad_x = grad(lambda x: torch.vmap(f)(x).sum())(x)
print(grad_x)
```
(which does error out correctly with this PR!)

soulitzer replied:
Oh ok, I see what you mean now. Nice!

zou3519 commented on test/test_autograd.py:
```diff
@@ -347,6 +347,37 @@ def test_custom_function_exception(self):
+            else:
+                with self.assertRaisesRegex(RuntimeError, "A input that has been returned as-is"):
+                    Func.apply(a)
+
```
- As a sanity check, does it make sense to add a test where we return an input as-is but do not save it for backward? I would expect this to not trigger the error.
- Could we add the same test, but run it under functorch.grad? It would be good to test that we're consistent between functorch and regular autograd.

soulitzer replied:
I think we already have the test in test_autograd, but sounds good on the functorch side.

zou3519 commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -339,6 +341,12 @@ optional_variable_list _process_backward_mode_ad(
         impl::rebase_history(var, {cdata, output_nr});
       }
     } else if (is_input) {
+      TORCH_CHECK(
```
Is it possible for (!is_differentiable && is_input) to be true? Is this the case where input does not require grad, so it is not ambiguous if we save the input or the input-returned-as-output for backward?

(If yes, then could we add a test for this to check that this case doesn't error out?)

soulitzer replied:
After testing with functorch some more, I realized that an error actually isn't actually raised in functorch unless the inner tensors wrapped by GradWrapper themselves require grad. I think its because to the perspective of the outer apply, the function never returned an input as-is (since the inner apply actually applied alias).

One way to fix that is actually to error in the !is_differentiable case as well, so we DO want to error out in this case. I think its reasonable to do this because the whether or not we pass in tensors that require grad, the code is still "bad" (the forward still returns input as-is, and setup_context still saves that tensor for backward), so we should make the user fix their code.

zou3519 replied:
> I think its because to the perspective of the outer apply, the function never returned an input as-is (since the inner apply actually applied alias).

Yeah, I think that is what is going on.

zou3519 reviewed (commented):
I added some questions about the autograd code (since I'm not very familiar with it) and also suggested some more test cases to convince ourselves that the fix is correct and doesn't affect other things. The code and approach looks good to me -- I wish the autograd code was less convoluted but what we're doing here makes sense.

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for forward or backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for forward or backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

zou3519 commented on test/functorch/test_eager_transforms.py:
```diff
@@ -1035,6 +1035,47 @@ def f(y, x):
+        def fn(x):
+            return A.apply(x.clone())
+
+        x = torch.tensor(2., device=device, requires_grad=inner_requires_grad)
+        if save_tensors in ("input", "output") and not mark_dirty:
+            with self.assertRaisesRegex(RuntimeError, "A input that has been returned as-is"):
+                grad(fn)(x)
+        else:
+            grad(fn)(x)
```
We should test running jvp as well, right?

soulitzer replied:
Yeah we could, but we don't think we have to. The error raising only cares about whether ctx.save_for_forward is called. It doesn't matter whether we actually doing jvp.

zou3519 replied:
Up to you, it's not a major point. My general philosophy is that if there isn't a test for it, someone will (unintentionally) find some way to break it. For example, we could decide to add an optimization where save_for_forward is not run unless the inputs are dual tensors and then somehow mess up the C++ logic and not notice it until later.

zou3519 commented on test/test_autograd.py:
```diff
@@ -347,6 +347,42 @@ def test_custom_function_exception(self):
             t3.sum().backward()
 
+    def test_custom_function_raises_if_input_returned_as_is_and_saved(self):
+        for save_inputs, mark_dirty, requires_grad, save_for_fw in product(*([(True, False)] * 4)):
```
why not use parametrize? it's not a functorch-only thing.

Also, nit: the test here and the test in test_eager_transforms are similar enough that it might be worth just shoving all of them in the same place, but up to you.

zou3519 commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -273,18 +273,28 @@ optional_variable_list _process_backward_mode_ad(
+    const std::unordered_set<at::TensorImpl*>& to_save_if_setup_context) {
   int num_outputs = raw_outputs.size();
 
+  std::string error_msg_input_returned_as_is =
```
Are we sure the compiler optimizes this well? `const char*` would give me more confidence

soulitzer replied:
OK, will update to const char* to be safe

zou3519 commented on torch/csrc/autograd/custom_function.cpp:
```diff
@@ -339,6 +349,7 @@ optional_variable_list _process_backward_mode_ad(
         impl::rebase_history(var, {cdata, output_nr});
       }
     } else if (is_input) {
+      TORCH_CHECK(!is_saved_and_setup_context, error_msg_input_returned_as_is)
```
Why don't we need a !is_modified check here?

soulitzer replied:
`if (is_modified) { ... } else if (is_input) { <we are here!> }`

zou3519 replied:
Oh lol the nesting got to me. Thanks for pointing that out!

zou3519 reviewed (approved):
LGTM, some minor comments

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for forward or backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge

soulitzer commented:
@pytorchbot revert -m "TestAutogradFunctionCUDA.test_function_returns_input_inner_requires_grad_True_save_for_vjp_save_tensors_output_mark_dirty_True_cuda leaks"

soulitzer commented:
@pytorchbot revert -m "TestAutogradFunctionCUDA.test_function_returns_input_inner_requires_grad_True_save_for_vjp_save_tensors_output_mark_dirty_True_cuda leaks" -c nosignal

zou3519 commented:
Does this PR cause a memory leak or is it that the memory-leak is pre-existing and we just added a test that exposes it?

soulitzer commented:
Preexisting

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for forward or backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented on test/functorch/test_eager_transforms.py:
```diff
@@ -1035,6 +1035,84 @@ def f(y, x):
+                self.assertTrue(a is b)
+            if not (mark_dirty and save_for == "vjp" and save_tensors in ("input", "output")):
+                # TODO(soulitzer): https://github.com/pytorch/pytorch/issues/97827
+                with fwAD.dual_level():
```
Filed an issue about the preexisting leak and updated the test just to skip this case for now.

soulitzer committed:
Update on "[BE][autograd Function] Raise an error if input is returned as-is and saved for forward or backward in setup_context"


Fixes https://github.com/pytorch/pytorch/issues/96887

We error out in BOTH the case when graph is created and when it is not created.

Still bc-breaking, but not as severe because we are limiting to the case where someone uses setup_context.

This makes setup_context and non-setup_context versions diverge in their behavior
- With the non-setup_context version, saved variables are assumed to have the grad_fn of the inputs. 
- But now with the setup_context version, we produce an error for this case.

cc ezyang gchanan

[ghstack-poisoned]

soulitzer commented:
@pytorchbot revert -m "Internally someone is rely on _wrap_outputs and we updated its signature" -c ghfirst

soulitzer commented:
Going to reopen this in a new non-ghstack PR so I can do codev

kit1980 commented:
> Internally someone is rely on _wrap_outputs

:-(
Another problem caused by users relying on private APIs..

albanD commented:
What? Someone relies on _wrap_outputs?? Do we even expose that function through our binary in any way?>

soulitzer commented:
Yeah it is TORCH_API to be fair lol
See D44552090 for the failure - It looks like someone reimplemented torch::autograd::Function::apply (so that they could reimplement activation checkpoint in cpp).
